pkg/controller/direct/kms/autokeyconfig_controller.go

```diff
@@ -0,0 +1,300 @@
+	if err != nil {
+		return err
+	}
+	status := &krm.KMSAutokeyConfigStatus{}
+	status.ObservedState = KMSAutokeyConfigStatusObservedState_FromProto(mapCtx, updated)
+	if mapCtx.Err() != nil {
+		return mapCtx.Err()
+	}
+	status.ExternalRef = a.id.AsExternalRef()
+	return setStatus(u, status)
+}
+
+func (a *Adapter) Update(ctx context.Context, updateOp *directbase.UpdateOperation) error {
+	u := updateOp.GetUnstructured()
+	log := klog.FromContext(ctx).WithName(ctrlName)
+	log.V(2).Info("updating AutokeyConfig", "name", a.id.FullyQualifiedName())
+	mapCtx := &direct.MapContext{}
+
+	desired := a.desired.DeepCopy()
+	resource := KMSAutokeyConfigSpec_ToProto(mapCtx, &desired.Spec)
+	if mapCtx.Err() != nil {
+		return mapCtx.Err()
+	}
+
+	updated, err := a.updateAutokeyConfig(ctx, resource)
+	if err != nil {
+		return err
+	}
+
+	status := &krm.KMSAutokeyConfigStatus{}
+	status.ObservedState = KMSAutokeyConfigObservedState_FromProto(mapCtx, updated)
+	if mapCtx.Err() != nil {
+		return mapCtx.Err()
+	}
+	return setStatus(u, status)
+}
+
+func (a *Adapter) updateAutokeyConfig(ctx context.Context, resource *kmspb.AutokeyConfig) (*kmspb.AutokeyConfig, error) {
+	log := klog.FromContext(ctx).WithName(ctrlName)
+	// to populate a.actual
+	isExist, err := a.Find(ctx)
+	if !isExist {
+		return nil, fmt.Errorf("updateAutokeyConfig failed as AutokeyConfig does not exist, name: %s", a.id.FullyQualifiedName())
+	}
+	if err != nil {
+		return nil, err
+	}
+	updateMask := &fieldmaskpb.FieldMask{}
+	if resource.KeyProject != "" && !reflect.DeepEqual(resource.KeyProject, a.actual.KeyProject) {
+		updateMask.Paths = append(updateMask.Paths, "key_project")
+	}
+
+	if len(updateMask.Paths) == 0 {
+		log.V(2).Info("no field needs update", "name", a.id.FullyQualifiedName())
+		return nil, nil
+	}
+	req := &kmspb.UpdateAutokeyConfigRequest{
+		UpdateMask:    updateMask,
+		AutokeyConfig: resource,
+	}
+	updated, err := a.gcpClient.UpdateAutokeyConfig(ctx, req)
+	if err != nil {
+		return nil, fmt.Errorf("updating AutokeyConfig %s: %w", a.id.FullyQualifiedName(), err)
+	}
+	log.V(2).Info("successfully updated AutokeyConfig", "name", a.id.FullyQualifiedName())
+	return updated, nil
+}
+
+func (a *Adapter) Export(ctx context.Context) (*unstructured.Unstructured, error) {
+	if a.actual == nil {
+		return nil, fmt.Errorf("Find() not called")
+	}
+	u := &unstructured.Unstructured{}
+
+	obj := &krm.KMSAutokeyConfig{}
+	mapCtx := &direct.MapContext{}
+	obj.Spec = direct.ValueOf(KMSAutokeyConfigSpec_FromProto(mapCtx, a.actual))
+	if mapCtx.Err() != nil {
+		return nil, mapCtx.Err()
+	}
+	obj.Spec.FolderRef = &refs.FolderRef{External: a.id.Parent.String()}
+	uObj, err := runtime.DefaultUnstructuredConverter.ToUnstructured(obj)
+	if err != nil {
+		return nil, err
+	}
+	u.Object = uObj
+	return u, nil
+}
+
+// Delete implements the Adapter interface.
+func (a *Adapter) Delete(ctx context.Context, deleteOp *directbase.DeleteOperation) (bool, error) {
```

Comment:
qq: Can KMS support unset a autokeyconfig, like changing the keyProject back to UNINITIALIZED?

Reply:
Yes, in the update operation if user sets `key_project` to empty then the state will be `UNITIALIZED`

Reply:
Thanks for the info. To reflect that in KCC, the update function would allow empty `keyProject`, and update the key_project` to UNINITIALIZED.

Reply:
yeah, `key_project` and `state` are two different fields in `autokeyConfig`.
The `state` field will be populated by the backend based on the value of `key_project`.
So as part of `update` call we will receive the right value of `state` and we dont need to handle the case in `update` api in controller